pkg/kubelet/metrics/metrics.go

```diff
@@ -736,6 +737,20 @@ var (
 			StabilityLevel: metrics.ALPHA,
 		},
 	)
+
+	// ContainerStartDuration tracks container start duration.
+	ContainerStartDuration = metrics.NewHistogramVec(
+		&metrics.HistogramOpts{
+			Subsystem:      KubeletSubsystem,
+			Name:           ContainerStartDurationKey,
+			Help:           "Duration in seconds to start a container.",
+			Buckets:        []float64{0.5, 1, 2, 3, 4, 5, 6, 8, 10, 20, 30, 45, 60, 120, 180, 240, 300, 360, 480, 600, 900, 1200, 1800, 2700, 3600},
+			StabilityLevel: metrics.ALPHA,
+		},
+		// container_type could be either "container", "init_container", or "ephemeral_container".
+		// namespace is the namespace of the pod that the container belongs to.
+		[]string{"container_type", "namespace"},
```

Comment:
1. Will we support a `sidecar` container as a special type of init_container?

2. I found that there are no histogram metrics using namespace as label. I think this is because there may be many pods from many namespaces in the kubelet lifecycle and it will make an unlimited size of metrics for this single one.

Reply:
> Will we support a sidecar container as a special type of init_container?

We cannot tell if a container is a sidecar or an init container here. If there's any way to tell it, we can/should certainly mark sidecar container durations with a separate label. cc @SergeyKanzhelev 


> I found that there are no histogram metrics using namespace as label. I think this is because there may be many pods from many namespaces in the kubelet lifecycle and it will make an unlimited size of metrics for this single one.

yep it is a valid concern. My hope was that the number of namespaces is smaller than that of pod names, and can be an acceptable metric field. But if not, I can remove that from the fields and only keep `container_type`.

Reply:
A namespace label can be very dangerous especially in testing clusters where test namespaces with generated names may be created which then also results in a lot of churn.

In this particular case, I don't really see how `namespace` could be a useful information. Container start time shouldn't depend on namespaces at all. I would understand more a `node` label but that still pretty huge in terms of cardinality on big clusters although far less than namespaces. But I don't think this particular metric justify this amount of cardinality.

@ruiwen-zhao could you perhaps share what could be potential reasons for container start latency?

Reply:
> My hope was that the number of namespaces is smaller than that of pod names, and can be an acceptable metric field

Even without taking into account cardinality concerns, is it really valuable to know that a particular pod in a particular namespace was slow to start? What are the action that one could take from that?
I find it far more valuable to know if the overall start is becoming slower and then maybe have metrics in the deeper layer of container startup that would give more info about what was slower than usual.